Fix slow queries in snapshot process

eileenmcnaughton opened the pull request:
SELECT COUNT(*) is slow to run on a large database but SELECT MAX(id) is quick. The snapshot check only needs to be very rough - so MAX(id) should be accurate enough.

This does make the message a little inaccurate - but my money says no-one will notice & it's not worth creating work for the translators to change

eileenmcnaughton committed:
Fix slow queries in snapshot process

SELECT COUNT(*) is slow to run on a large database but SELECT MAX(id) is
quick. The snapshot check only needs to be very rough - so MAX(id) should
be accurate enough.

This does make the message a little inaccurate - but my money says no-one
will notice & it's not worth creating work for the translators to change

seamuslee001 commented:
@eileenmcnaughton @return doc comment specified, but function has no return statement

demeritcowboy commented on CRM/Upgrade/Snapshot.php:
```diff
@@ -57,15 +60,16 @@ public static function getActivationIssues(): array {
         }
-        catch (\Exception $e) {
-          $count = 0;
+        catch (CRM_Core_Exception $e) {
```
These are usually PEAR_Exceptions

eileenmcnaughton replied:
@demeritcowboy thrown from the `CRM_Core_DAO::executeQuery` ?

demeritcowboy replied:
Yes.

You can simulate it with something like `cv ev 'try { CRM_Core_DAO::singleValueQuery("SELECT begargle FROM civicrm_contact LIMIT 1"); } catch (\Exception $e) { echo get_class($e); }'`

eileenmcnaughton replied:
ok - so on one hand I should possibly fix to catch the PEAR_EXCEPTION - but I'm also  inclined to add a new exception `CRM_DatabaseQuery_Exception` that extends `CRM_Core_Exception` - that way we can a) catch all civi exceptions within civi & also know that it is a db exception

totten commented:
Needs a style fixup.

Would [information_schema.tables.table_rows](https://dba.stackexchange.com/a/184746) be any better or worse than `MAX(id)`?

eileenmcnaughton commented:
@totten my gut says it would be worse & that there is a reason I avoid querying the information schema for that - but it is a number of years since I last tried

totten committed:
(NFC) CRM/Upgrade/Snapshot.php - Work-around for bug in linter

The linter for `@return` has a bug; it expects to see a `return`
and doesn't recognize the `yield` notation (which is radically new,
having been added in PHP 5.5 circa 2013).

totten commented:
> my gut says it would be worse & that there is a reason I avoid querying the information schema for that -

Well, I avoid `information_schema` because it feels alien. 👽 

There may be some edge-case trade-offs in each direction, but yeah - it's not really worth addressing speculatively. We can use the current formulation and adjust if anyone reports an issue.

Anywho, merge-on-pass.

eileenmcnaughton commented on CRM/Upgrade/Snapshot.php:
```diff
@@ -133,8 +133,6 @@ public static function createTableName(string $owner, string $version, string $n
    * @param string $name
    * @param \CRM_Utils_SQL_Select $select
-   *
-   * @return iterable
```
If this needs to go then the typehint is wrong too

totten replied:
No, the type-hint is correct. The `@return` was correct, too. As discussed in the git commit message, the problem is a bug in the phpcs rule. But I didn't want your entire PR blocked on figuring out a bugfix for phpcs, and it's preferrable to have *some* indication of the return type.
